vavr/src/main/java/io/vavr/control/Either.java

```diff
@@ -414,6 +414,44 @@ default void orElseRun(Consumer<? super L> action) {
         return isLeft() || predicate.test(get()) ? Option.some(this) : Option.none();
     }
 
+    /**
+     * Filters this right-biased {@code Either} by testing a predicate.
+     * If the {@code Either} is a Right and the predicate doesn't match, the
+     * {@code Either} will be turned into a {@code Left} with contents computed by applying
+     * the filterVal function to the {@code Either} value.
+     * <p>
+     *
+     * <pre>{@code
+     * import static io.vavr.API.*;
+     *
+     * class Example {
+     *
+     *     // = Left("bad: a")
+     *     Right("a").filterEither(i -> false, val -> "bad: " + val);
+     *
+     *     // = Right("a")
+     *     Right("a").filterEither(i -> true, val -> "bad: " + val);
+     *
+     * }
+     * }</pre>
+     *
+     * @param predicate A predicate
+     * @return an {@code Either} instance
+     * @throws NullPointerException if {@code predicate} is null
+     */
+    default Either<L,R> filterEither(Predicate<? super R> predicate,
+                                     Function<? super R, ? extends L> filterVal) {
```

Comment:
Let's rename `filterEither` to `filterOrElse` in order to align a bit more to Scala (see [Scala API](https://www.scala-lang.org/api/2.12.5/scala/util/Either.html#filterOrElse[A1>:A](p:B=>Boolean,zero:=>A1):scala.util.Either[A1,B])).

Maybe we should rename `filterVal` also to `zero`, it puts more emphasis that it is used in the empty case.

Scala does have a lazy zero provider `zero: => L`. But I think your choice is better in terms of flexibility. One does not have to use the right value in order to derive a left value. In other cases we use additional overloads (_Supplier&lt;L> zero_ and _L zero_), but I'm not a big fan of adding such a verbosity to the API.